chore: Update dockerfile for dev and prod

TimVosch opened the pull request:
This is a side effect of my WIP PR that allows switching between DomPDF and Gotenberg for PDF generation.

I needed an easier way to build a production ready image for my Kubernetes environment. The Dockerfile at invoiceshelf-docker repository does a git clone from this repository which complicates things when working from a local env / different git repository.

I've opted to use the PHP and Laravel optimized serversideup base container. This greatly simplifies the docker build process and it does some proper layer caching, making builds faster and concurrently using buildx.

The Dockerfile has a version of imagick pinned due to some build issues with the latest imagick version on arm64, when cross building using:

```
docker buildx build --platform=linux/amd64,linux/arm64 -t <reponame> --push .
```

I've also added the network alias of `invoiceshelf.test` to the invoiceshelf container, so that other services can resolve that domain to the correct container.

I've been running this image for a few days in dev now using sqlite. Haven't tested psql and msql yet.

TimVosch committed:
chore: update dockerfile and dev env

TimVosch committed:
chore(dockerfile): fix user/group id args

TimVosch committed:
chore(docker): use php-fpm w/ separate nginx

TimVosch commented:
I'm testing it on my "production" environment and got some issues with the conf and data volumes. I am just going to move this to a draft as well until I've sorted that out.

rihards-simanovics commented:
apologies @TimVosch, got trigger happy 😅

TimVosch committed:
chore(docker): add nginx image w/ static files

TimVosch commented:
To give some context for some of the choices, I will be running InvoiceShelf on my personal Kubernetes cluster. I am doing my best to support a most "stateless" form of InvoiceShelf. And support Kubernetes in a "conventional" way.

I've added a separate `nginx.Dockerfile` which builds the static assets and creates an NGINX image. This image uses a environment variable to point to the PHP-FPM server.

Some points I'd like to have your opinion on:

- The original entrypoint script links a `.env` from a `/conf` mount to the app folder, and it copies a sqlite database file from the mount to the app/database folder. 
  - What is the benefit of having a .env file in a containerized environment? I get it for a development environment, but that will be mounted in the development docker compose scrips.
  - Instead of copying the database around, can we just mount it elsewhere and use the corresponding DB_... environment variable to point to it?

- The original entrypoint script also merges the /data and storage folder, as shown below. Why not just mount the data mount at the storage folder directly?

```sh
echo "**** Link storage ****"
[ ! -L /var/www/html/InvoiceShelf/storage ] && \
	cp -r /var/www/html/InvoiceShelf/storage/* /data && \
	rm -r /var/www/html/InvoiceShelf/storage && \
	ln -s /data /var/www/html/InvoiceShelf/storage

echo "**** Expose storage ****"
[ ! -L /var/www/html/InvoiceShelf/public/storage ] && \
	ln -s /data/app/public /var/www/html/InvoiceShelf/public/storage
```

TimVosch committed:
chore(docker): build vite resources only once, bump vite minor version,
add watch yarn command.
By using --buildplatform tag in the dockerfile we can have the vite step
be built only on the host platform, which significantly speeds it up.
This is possible since the build assets aren't platform dependant.

TimVosch commented on vite.config.js:
```diff
@@ -1,42 +1,48 @@
-            $images: resolve(__dirname, './resources/static/img')
-        },
-        extensions: ['.js', '.ts', '.jsx', '.tsx', '.json', '.vue', '.mjs']
+  server: {
+    host: 'invoiceshelf.test',
+    hmr: {
+      host: 'invoiceshelf.test',
+    }
+  },
```
Only this changed, but I formatted the file on accident. 

This line makessure that the `vite build --watch` (called by `yarn dev`) command works on the http://invoiceshelf.test domain. 

The watch command is super responsive and very nice during development ;)

TimVosch commented:
@gdarko I think you can review it. 
Just a headsup that the dockerfile is production ready, but has some different settings as opposed to the production dockerfile at https://github.com/InvoiceShelf/docker !
See my previous comment for info on that.

TimVosch commented:
I could spend some time to have the .dev docker compose files also initiate the `vite build --watch`, to make development easier

rihards-simanovics commented:
Hey @TimVosch, sorry for the long wait, I am looking through the old PRs here. Is this PR final and ready for review?

faabiosr commented:
@TimVosch What's the point of moving the `Dockerfiles` from `.dev`? At least for me, the previous one had everything in one place, making it easier to understand.

TimVosch committed:
Move dockerfiles to .dev

TimVosch commented:
I rebased this branch on master.

@faabiosr no reason why the dockerfiles were in root, I moved it them to .dev/

@rihards-simanovics I think @gdarko had some feedback on this in (https://github.com/InvoiceShelf/InvoiceShelf/pull/184#pullrequestreview-2588314581)

It has been a while but I think these were some of the benefits:
1. The Dockerfile uses multiple stages allowing for a development environment (as .dev is supposed to have) and a production build from the same place.
2. The Dockerfile has a production stage which is - in my opinion - way less complicated than that in https://github.com/InvoiceShelf/docker and is representative of the current commit / tag it's being build on.
3. Less important, but the dockerfile is suitable for cross-architecture image i.e. building on amd64 for arm64 will build the static files on the host machine (which is way faster) and copy them over to the arm64 image. (I run an arm64 cluster in production)
4. Changes in `vite.config.js` and package.json (just `vite build --watch`) allow laravel and vite to actually work together to provide HMR. Which make changes in the Vue files appear almost instantly in the browser instead of waiting for an entire new vite build to rollup.

Try it out and see if it works for you.

gdarko commented:
Thank you @TimVosch - This is now merged.
